rfc/0075-faster-failover-and-configuration-push.md

```diff
@@ -0,0 +1,315 @@
+this particular node instead.
+
+### Enhancements in Handling the `NotMyVbucket` Status
+
+Combination of `DedupeNotMyVbucketClustermap` and `ClustermapChangeNotificationBrief` allows to save traffic by not
+sending configuration, if SDK already seen the same revision, and also sends only pair of `Epoch`/`Revision`. So it is
+up to SDK to initiate configuration update once the non-empty payload returned along with `NotMyVbucket` status code.
+
+Several modifications are required in the SDK:
+1. The retry orchestrator should be able to retry an operation based on configuration updates rather than the timer signal.
+2. The configuration monitor should have the ability to throttle configuration requests due to the following reasons:
+   1. During rebalance, multiple operations may return a `NotMyVbucket` status, triggering a configuration refresh.
+   2. Since `ClustermapChangeNotificationBrief` will cause all connections to subscribe to updates and receive them, it is
+      necessary to account for potential high volumes of updates.
+
+Below is a diagram that illustrates an example of the SDK workflow, where the GET request is waiting for the arrival of
+a new configuration.
+
+```mermaid
+sequenceDiagram
+    autonumber
+    conn_1->>+kv_node_1: get("foo", vb=115)
+    kv_node_1->>-conn_1: NotMyVbucket(epoch=1, rev=11)
+    conn_1-->>+retry_orchestrator: pending(get, "foo", epoch=1, rev=11)
+    retry_orchestrator-->retry_orchestrator: put operation to wating queue
+    conn_1-->>+config_monitor: refresh configuration
+    config_monitor-->config_monitor: wait to throttle config requests
+    config_monitor->>+conn_2: get_config()
+    conn_2->>+kv_node_2: get_config()
+    kv_node_2->>-conn_2: configuration(epoch=1, rev=11)
+    conn_2->>-config_monitor: apply new configuration
+    config_monitor->>retry_orchestrator: purge waiting queue(epoch=1, rev=11)
+    retry_orchestrator->>conn_1: retry get("foo")
+    conn_1->>+kv_node_2: get("foo", vb=115)
+    kv_node_2->>-conn_1: Success()
+```
+
+# Language Specifics
+
+## Feature Checklist
+
+1. `GetClusterConfigWithKnownVersion` (`0x1d`). The SDK should always supply current configuration version if the
+   connection has acknowledged feature flag.
+
+2. `DedupeNotMyVbucketClustermap` (`0x1e`). The SDK should be ready that the KV engine will not repeat configuration
+   payload if it already been sent to the socket by any means (`NotMyVbucket` status, `ClustermapChangeNotification`,
+   `GetClusterConfig`).
+
+3. Out-of-Order Execution. `Duplex` (`0x0c`) feature should be always negotiated in HELLO.
+
+4. `ClustermapChangeNotificationBrief` (`0x1f`). The SDK should always subscribe for configuration notifications, if the
+   server supports it, and fallback to polling if it does not.
+
+5. SDK should not emit configuration refresh request if there is one already in-flight. This should be independent of
```

Comment:
There is a race that can occur here.  If we have an in-flight request for a configuration when we receive a new event that requires another configuration, it's possible that the server will reply to the first request with a response that predates the later event.  I think we need to consider the revision information from the event that triggered the refresh request such that if the original configuration request that was in progress returns an older configuration, that we dispatch another config request.

Reply:
Right, it makes sense. So I will rephrase this.